lightly/api/download.py

```diff
@@ -0,0 +1,192 @@
+import concurrent.futures
+import os
+import pathlib
+import shutil
+import threading
+import warnings
+from concurrent.futures import ThreadPoolExecutor
+from typing import List, Iterable, Union, Tuple, Dict
+
+import av
+import requests
+import tqdm
+from lightly.api import utils
+import PIL
+
+
+def download_image(url: str, session: requests.Session = None) -> PIL.Image.Image:
+    """Downloads an image from an url.
+
+    Args:
+        url: 
+            Url where image is stored.
+        session: 
+            Make request using the given session.
+
+    Returns:
+        The downloaded image.
+
+    """
+    req = requests if session is None else session
+    response = req.get(url, stream=True)
+    return PIL.Image.open(response.raw)
+
+
+def download_all_video_frames(
+    url: str, as_pil_image: int = True
+) -> Iterable[Union[PIL.Image.Image, av.VideoFrame]]:
+    """Lazily retrieves all frames from a video.
+
+    Args:
+        url: 
+            Url where video is stored.
+        as_pil_image: 
+            Whether to return the frame as PIL.Image.
+
+    Returns:
+        A generator that loads and returns a single frame per step.
+
+    """
+    container = av.open(url)
+    stream = container.streams.video[0]
+    stream.thread_type = 'AUTO'
+    for frame in container.decode(stream):
+        if as_pil_image:
+            yield frame.to_image()
+        else:
+            yield frame
+    container.close()
+
+
+def download_video_frame(
```

Comment:
why did you change from access by index to access by timestamp?

Reply:
You cannot seek to a specific frame index in pyav. The problem is that videos do not necessarily store frames in fixed intervals. You would have to read the complete video first to map the frame index to a timestamp. Torchvision also avoids frame indices and uses timestamps instead.